datalad_dataverse/remote.py

```diff
@@ -92,111 +136,311 @@ def api(self):
+                                   (v.get('versionNumber') or sys.maxsize,
+                                    v.get('versionMinorNumber') or sys.maxsize),
+                               reverse=True)
+
+        return self._dataset
+
+    @property
+    def files_all(self):
+        """Cache of files available from dataverse.
+
+        For quick lookup and deduplication, this is a dict {id: remote_path}
+        where remote_path is a `Path` build from the file's 'directoryLabel'
+        (if any) and 'filename'.
+        Currently, this is all we need, hence don't bother carrying and updating
+        all other metadata.
+        """
+
+        if self._files_all is None:
+            self._files_all = {
+                f['dataFile']['id']: Path(f.get('directoryLabel', '')) / f['dataFile']['filename']
+                for f_l in [version['files'] for version in self.dataset]
+                for f in f_l}
+
+        return self._files_all
+
+    @property
+    def files_latest(self):
+        """Cache of files in the latest version of the dataverse dataset.
+
+        This refers to the DRAFT version (if there is any) or the latest
+        published version otherwise. That's the version pushes go into. Hence,
+        this is needed to determine whether we need to replace a file on
+        upload, while the complete list in `self.files_all` is relevant for key
+        retrieval (which can come from earlier dataverse versions of the
+        dataset).
+        """
+
+        if self._files_latest is None:
+            # Latest version in self.dataset is first entry.
+            self._files_latest = {
+                f['dataFile']['id']: Path(f.get('directoryLabel', '')) / f['dataFile']['filename']
+                for f in self.dataset[0]['files']}
+
+        return self._files_latest
+
+    def remove_from_filelist(self, id):
+        """Update self.filelist after removal"""
+        if self._files_all is None:
+            self.message(f"FAILED Removing {id} from file list failed. "
+                         f"'_filelist' does not exist.",
+                         type='debug')
+            return  # not yet a reason to crash
+        self._files_all.pop(id)
+        try:
+            self._files_latest.pop(id)
+        except KeyError:
+            # This is alright, we removed from _all, it may not have been in
+            # latest, just calling here to be sure we are consistent between the
+            # two.
+            pass
+
+    def add_to_filelist(self, d):
+        """Update self.filelist after upload
+
+        d: dict
+          dataverse description dict of the file; this dict is in the list
+          'data.files' of the response to a successful upload
+        """
+        entry = (d['dataFile']['id'],
+                 Path(d.get('directoryLabel', '')) /
+                    d['dataFile']['filename']
+                 )
+        self._files_all[entry[0]] = entry[1]
+        self._files_latest[entry[0]] = entry[1]
```

Comment:
Another paired modification. Still not sure what the purpose of `files_latest` is, but it dawns on me that `files_all` should be a mapping `id: (path, bool=in_latest)`.